File: apps/api/app/api/v1/community.py
from app.core.config import Settings, get_settings
from app.core.database import get_connection
from app.schemas.community import (
    CommunityAnswer,
    CommunityAnswerCreate,
    CommunityQuestion,
    CommunityQuestionCreate,
    CommunityQuestionListResponse,
    CommunityVoteCreate,
    ConsensusSummary,
)
from app.schemas.data_error_reports import DataErrorReport, DataErrorReportCreate
from app.schemas.user_story_ratings import UserStoryRating, UserStoryRatingCreate
from app.services import (
    community as community_service,
    data_error_reports as data_error_reports_service,
    user_story_ratings as user_story_ratings_service,
)
from fastapi import APIRouter, Depends, Query
from psycopg import Connection
from typing import Annotated, Any


router = APIRouter(prefix="/community", tags=["community"])


@router.get("/questions", response_model=CommunityQuestionListResponse)
def list_questions(
    connection: Annotated[Connection[Any], Depends(get_connection)],
    country_slug: str | None = Query(None),
    topic: str | None = Query(None),
    limit: int = Query(50, ge=1, le=200),
) -> CommunityQuestionListResponse:
    items = community_service.list_public_questions(
        connection, country_slug=country_slug, topic=topic, limit=limit
    )
    return CommunityQuestionListResponse(items=items, total=len(items))


@router.get(
    "/questions/{question_id}",
    response_model=CommunityQuestion,
    responses={404: {"description": "Not found"}},
)
def get_question(
    question_id: str,
    connection: Annotated[Connection[Any], Depends(get_connection)],
) -> dict[str, Any]:
    return community_service.get_public_question(connection, question_id)


@router.post("/questions", response_model=CommunityQuestion, status_code=201)
def create_question(
    payload: CommunityQuestionCreate,
    connection: Annotated[Connection[Any], Depends(get_connection)],
    settings: Annotated[Settings, Depends(get_settings)],
) -> dict[str, Any]:
    row = community_service.submit_question(connection, settings, payload)
    connection.commit()
    return row


@router.get(
    "/questions/{question_id}/answers",
    response_model=list[CommunityAnswer],
)
def list_answers(
    question_id: str,
    connection: Annotated[Connection[Any], Depends(get_connection)],
) -> list[dict[str, Any]]:
    return community_service.list_public_answers(connection, question_id)


@router.post(
    "/questions/{question_id}/answers",
    response_model=CommunityAnswer,
    status_code=201,
    responses={404: {"description": "Not found"}},
)
def create_answer(
    question_id: str,
    payload: CommunityAnswerCreate,
    connection: Annotated[Connection[Any], Depends(get_connection)],
    settings: Annotated[Settings, Depends(get_settings)],
) -> dict[str, Any]:
    row = community_service.submit_answer(connection, settings, question_id, payload)
    connection.commit()
    return row


@router.post(
    "/answers/{answer_id}/votes",
    response_model=ConsensusSummary,
    responses={404: {"description": "Not found"}},
)
def create_vote(
    answer_id: str,
    payload: CommunityVoteCreate,
    connection: Annotated[Connection[Any], Depends(get_connection)],
    settings: Annotated[Settings, Depends(get_settings)],
) -> ConsensusSummary:
    result = community_service.submit_vote(connection, settings, answer_id, payload)
    connection.commit()
    return result


@router.post("/data-error-reports", response_model=DataErrorReport, status_code=201)
def create_data_error_report(
    payload: DataErrorReportCreate,
    connection: Annotated[Connection[Any], Depends(get_connection)],
    settings: Annotated[Settings, Depends(get_settings)],
) -> dict[str, Any]:
    row = data_error_reports_service.submit_data_error_report(
        connection, settings, payload
    )
    connection.commit()
    return row


@router.post("/user-story-ratings", response_model=UserStoryRating, status_code=201)
def create_user_story_rating(
    payload: UserStoryRatingCreate,
    connection: Annotated[Connection[Any], Depends(get_connection)],
    settings: Annotated[Settings, Depends(get_settings)],
) -> dict[str, Any]:
    row = user_story_ratings_service.submit_user_story_rating(
        connection, settings, payload
    )
    connection.commit()
    return row

File: apps/web/src/shared/api/community.ts
import type { components } from "@country-decision-atlas/contracts/generated/types";

import { csrfHeaders } from "../auth/session";
import { apiGet, apiPatch, apiPost, queryString } from "./http";

export type CommunityQuestion = components["schemas"]["CommunityQuestion"];
export type CommunityQuestionCreate =
  components["schemas"]["CommunityQuestionCreate"];
export type CommunityQuestionListResponse =
  components["schemas"]["CommunityQuestionListResponse"];
export type CommunityAnswer = components["schemas"]["CommunityAnswer"];
export type CommunityAnswerCreate =
  components["schemas"]["CommunityAnswerCreate"];
export type CommunityVoteCreate = components["schemas"]["CommunityVoteCreate"];
export type ConsensusSummary = components["schemas"]["ConsensusSummary"];
export type DataErrorReport = components["schemas"]["DataErrorReport"];
export type DataErrorReportCreate =
  components["schemas"]["DataErrorReportCreate"];
export type UserStoryRating = components["schemas"]["UserStoryRating"];
export type UserStoryRatingCreate =
  components["schemas"]["UserStoryRatingCreate"];
export type CommunityStatusUpdateRequest =
  components["schemas"]["CommunityStatusUpdateRequest"];
export type DataErrorReportStatusUpdateRequest =
  components["schemas"]["DataErrorReportStatusUpdateRequest"];
export type UserStoryRatingStatusUpdateRequest =
  components["schemas"]["UserStoryRatingStatusUpdateRequest"];

export function listCommunityQuestions(params: {
  country_slug?: string;
  topic?: string;
  limit?: number;
}): Promise<CommunityQuestionListResponse> {
  return apiGet<CommunityQuestionListResponse>(
    `/api/v1/community/questions${queryString(params)}`,
  );
}

export function listCommunityAnswers(
  questionId: string,
): Promise<CommunityAnswer[]> {
  return apiGet<CommunityAnswer[]>(
    `/api/v1/community/questions/${questionId}/answers`,
  );
}

export function createCommunityQuestion(
  payload: CommunityQuestionCreate,
): Promise<CommunityQuestion> {
  return apiPost<CommunityQuestion, CommunityQuestionCreate>(
    "/api/v1/community/questions",
    payload,
  );
}

export function createCommunityAnswer(
  questionId: string,
  payload: CommunityAnswerCreate,
): Promise<CommunityAnswer> {
  return apiPost<CommunityAnswer, CommunityAnswerCreate>(
    `/api/v1/community/questions/${questionId}/answers`,
    payload,
  );
}

export function voteCommunityAnswer(
  answerId: string,
  payload: CommunityVoteCreate,
): Promise<ConsensusSummary> {
  return apiPost<ConsensusSummary, CommunityVoteCreate>(
    `/api/v1/community/answers/${answerId}/votes`,
    payload,
  );
}

export function createDataErrorReport(
  payload: DataErrorReportCreate,
): Promise<DataErrorReport> {
  return apiPost<DataErrorReport, DataErrorReportCreate>(
    "/api/v1/community/data-error-reports",
    payload,
  );
}

export function createUserStoryRating(
  payload: UserStoryRatingCreate,
): Promise<UserStoryRating> {
  return apiPost<UserStoryRating, UserStoryRatingCreate>(
    "/api/v1/community/user-story-ratings",
    payload,
  );
}

export function listAdminCommunityQuestions(
  status?: string,
): Promise<CommunityQuestion[]> {
  return apiGet<CommunityQuestion[]>(
    `/api/v1/admin/community/questions${queryString({ status })}`,
    { headers: csrfHeaders() },
  );
}

export function updateCommunityQuestionStatus(
  questionId: string,
  payload: CommunityStatusUpdateRequest,
): Promise<CommunityQuestion> {
  return apiPatch<CommunityQuestion, CommunityStatusUpdateRequest>(
    `/api/v1/admin/community/questions/${questionId}/status`,
    payload,
    { headers: csrfHeaders() },
  );
}

export function listAdminCommunityAnswers(
  status?: string,
): Promise<CommunityAnswer[]> {
  return apiGet<CommunityAnswer[]>(
    `/api/v1/admin/community/answers${queryString({ status })}`,
    { headers: csrfHeaders() },
  );
}

export function updateCommunityAnswerStatus(
  answerId: string,
  payload: CommunityStatusUpdateRequest,
): Promise<CommunityAnswer> {
  return apiPatch<CommunityAnswer, CommunityStatusUpdateRequest>(
    `/api/v1/admin/community/answers/${answerId}/status`,
    payload,
    { headers: csrfHeaders() },
  );
}

export function listAdminDataErrorReports(
  status?: string,
): Promise<DataErrorReport[]> {
  return apiGet<DataErrorReport[]>(
    `/api/v1/admin/community/data-error-reports${queryString({ status })}`,
    { headers: csrfHeaders() },
  );
}

export function updateDataErrorReportStatus(
  reportId: string,
  payload: DataErrorReportStatusUpdateRequest,
): Promise<DataErrorReport> {
  return apiPatch<DataErrorReport, DataErrorReportStatusUpdateRequest>(
    `/api/v1/admin/community/data-error-reports/${reportId}/status`,
    payload,
    { headers: csrfHeaders() },
  );
}

export function listAdminUserStoryRatings(
  status?: string,
): Promise<UserStoryRating[]> {
  return apiGet<UserStoryRating[]>(
    `/api/v1/admin/community/user-story-ratings${queryString({ status })}`,
    { headers: csrfHeaders() },
  );
}

export function updateUserStoryRatingStatus(
  ratingId: string,
  payload: UserStoryRatingStatusUpdateRequest,
): Promise<UserStoryRating> {
  return apiPatch<UserStoryRating, UserStoryRatingStatusUpdateRequest>(
    `/api/v1/admin/community/user-story-ratings/${ratingId}/status`,
    payload,
    { headers: csrfHeaders() },
  );
}

export const communityApi = {
  listCommunityQuestions,
  listCommunityAnswers,
  createCommunityQuestion,
  createCommunityAnswer,
  voteCommunityAnswer,
  createDataErrorReport,
  createUserStoryRating,
  listAdminCommunityQuestions,
  updateCommunityQuestionStatus,
  listAdminCommunityAnswers,
  updateCommunityAnswerStatus,
  listAdminDataErrorReports,
  updateDataErrorReportStatus,
  listAdminUserStoryRatings,
  updateUserStoryRatingStatus,
};
